### ISSUE TITLE
Location of error message for uninitialized fields

### ISSUE THREAD

**Issue by mernst** [jspecify/jspecify#171]
File `UninitializedField`.java contains a `@NonNull` field that is not initialized.
The error message appears on the field.

In this particular test case, this is a natural place for the error message.  It is less natural if there is a user-defined constructor.

Suppose that the class contains 10 constructors, 9 of which initialize the variable and one of which does not.  In this case, I think the error message should be at the one incorrect constructor (this is probably the location of the error) rather than at the field.

Please add a test case to show where the error should be issued, if there is a user-defined constructor that does not initialize a `@NonNull` field.

**Comment by cpovirk** [comment]
On the one hand, I don't think we strictly _need_ to specify where the "mismatch" happens. And I would feel bad if we picked one location for our sample and then a tool felt compelled to go to a lot of trouble to match.

On the other hand, it would be nice to have such a sample (especially because of the 10-9 example you give). And that requires us to pick a location (unless we put the constructor and the field declaration on the same line ;)). I like the argument in favor of the constructor in that case, and I wouldn't mind nudging tool authors in that direction.

I'll try to get around to confirming where the Checker Framework reports it (on the `Foo()` line of the constructor?) and then add such a sample. Or certainly feel free to send another PR.

**Comment by mernst** [comment]
After https://github.com/typetools/checker-framework/pull/4259, the Nullness Checker's rule will be:

 * for static fields or the default constructor, issue the warning on the field.
 * for any user-written constructor, issue the warning on the field.

(This is a change the Checker Framework to accommodate the locations mandated by JSpecify tests, for static fields and when there is a default constructor.)

**Comment by cpovirk** [comment]
Right, when we write samples, we have to pick a location to put the comment on. (Again, though, we do not intend to mandate an error, nor mandate the location for an error if the case that a tool chooses to report one.) The unfortunate part is that that provides us a motivation to _not_ write samples: Once we pick a location, tool authors may find that to be reason to change their behavior -- either because they are concerned that people will see the location as mandated or because it simplifies their testing. And it's not entirely clear that we want them to put them in that position.

(Again, though, I do see the appeal of the specific behavior you've laid out for each of declared constructors, default constructors, and static fields. And assuming that I eventually write the sample, I'll follow that.)
